library/src/G4LatticeLogical.cc

```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec) {
+      
+  // Find the rotation matrix elements. 
+  // The rows correspond to the rotated axis direction (first row is vx,vy,vz ; second row is a,b,c and third row is d,e,f)
+    
+  // The convention used does not affect HV-transformation. For simplicity, we choose the following conditions :
+  // The rotated y axis must stay in the X-Y plane (c=0)
+  // The rotate Z axis must have positive z component (f>0) 
+  // The three rotated axis should stay normalized : vx^2 + vy^2 + vz^2 = 1 ; a^2 + b^2 = 1 ; d^2 + e^2 + f^2 = 1
+  // And be orthogonal : ad + be = 0 ; avx + bvy = 0 ; dvx + evy + fvz = 0
+  // It must be a right hand coordinate system : (vx,vy,vz) X (a,b,c) = (d,e,f) : d=-vzb ; e=vza ; f=vxb - vya
+    
+  // We first compute f since it needs to be positive and will therefore clear out signs ambiguity 
+  // d^2 + e^2 + f^2 = 1
+  // f^2 = 1 - d^2 - e^2
+  // f^2 = 1 - vz^2 b^2 - vz^2 a^2
+  // f^2 = 1 - vz^2 (b^2 + a^2)
+  // f^2 = 1 - vz^2
+  // f = + sqrt(1 - vz^2)
+    
+    
+  // Then we compute special cases first (vx and vy=0, vx=0 but not vy and vy=0 but not vx).
+    
+  // We start with vx= and vy=0
+  // e=0 and there is only two possible rotation matrix since vz=1 or -1. We manually set the elements to :
+  // b=1 
+  // a=0
+    
+  // If vx=0 but not vy 
+  // a = -e/vy = -sqrt(1-vz^2)/vy = -sqrt(vy^2)/vy = -1 for valleys and 1 for anti-valleys
+  // b=0
+
+  // If vy=0 but not vx 
+  // b = e/vx = sqrt(1-vz^2)/vx = sqrt(vx^2)/vx = 1 for valleys and -1 for anti-valleys
+  // a=0
+    
+    
+  // We go on with "general case" where  vx=!0 and vy!=0
+    
+  // We start by computing b (if vx=!0 and vy!=0) : 
+  // f = vxb - vya
+  // f = vxb - vy (-bvy/vx)
+  // f = b(vx + vy^2/vx)
+  // f = b/vx(vx^2 + vy^2)
+  // b = fvx/(vx^2 + vy^2)
+  // b = fvx/(1 - vz^2)
+  // b = fvx/f^2
+  // b = vx/f
+    
+  // Then we compute a (if vx=!0 and vx!=0) : 
+  // a = -bvy/vx
+  // a = -vxvy/e/vx
+  // a = -vy/e
+    
+    
+  // Finally, we compute d and e which is the same for every case :
+  // d=-vzb
+  // e=vza
+    
+    
+  double vx=valleyDirVec.x();
```

Comment:
All of this code needs to be consolidated to make better use of G4ThreeVector and G4RotationMatrix functionality.

Reply:
This still needs to be done.  If you're not sure how to do it, please say so here in these comments so we can work together to resolve.  Using the vector manipulations built into G4ThreeVector and G4RotationMatrix should make the code much more self-documenting.

Reply:
I made some changes by using a tempvect(), .unit() and G4Rep3x3, I'm not sure how to consolidate that more. I like using a double for vx=tempvec().x() because we are using vx a few times in the lines below. Maybe I should change that?

Reply:
Let's find time to go through together (maybe using your slides rather than the code directly) exactly what's going on in each stage.  My suspicion is that you "should" be able to build the rotation matrix similar to how Euler angles work, by computing a set of "rotate about X", "rotate about Z" actions based on the direction vector, rather than doing the algebra "by hand" for the nine matrix components.

However, I may be falling into the "round-off trap."  That is, if we do what I'm thinking, maybe we end up with a less precise matrix than if we consolidate the underlying arithmetic the way you are doing.